### PR TITLE
Add window channel

### PR DESCRIPTION
**Pull request by HakkyuKim**
The use case for setting window geometry at runtime is explained [here](https://github.sec.samsung.net/f-project/system_ui_app/issues/38) (internal repo).

Test app: https://github.com/HakkyuKim/flutter_tizen_window_resize_test_app

## TODO
- [x] Disable for evas_gl renderer.
- [x] Investigate why it works for TV and wearable emulators but not for mobile.
    Can't figure out why, decided not to support in this PR since it's not important compared to other tasks.
- [x] Investigate why x and y position are not set correctly.
    Must also call `ecore_wl2_window_position_set`.
- [x] ~~Test with rotation.~~
    I'll write a "remaining issues" in the issue tab.
- [x] Test with top_level window, #191.
- [x] Formatting.
- [x] ~~Check dynamic type of `TizenRenderer` in `WindowChannel::HandleMethodCall()`.~~ Use conditional compilation.

### CODE REVIEW

File: shell/platform/tizen/flutter_tizen_engine.cc [hunk 1]
```diff
@@ -245,7 +245,16 @@ bool FlutterTizenEngine::RunEngine(const char* entrypoint) {
...
   app_control_channel_ = std::make_unique<AppControlChannel>(
       internal_plugin_registrar_->messenger());
-#endif
+  if (IsHeaded()) {
+#ifdef TIZEN_RENDERER_EVAS_GL
+    window_channel_ = std::make_unique<WindowChannel>(
+        internal_plugin_registrar_->messenger(), renderer_.get());
+#else
+    window_channel_ = std::make_unique<WindowChannel>(
+        internal_plugin_registrar_->messenger(), renderer_.get(), this);
```
Comment (swift-kim) [inline shell/platform/tizen/flutter_tizen_engine.cc:254]: Please move this line to between line 275-276.

File: shell/platform/tizen/channels/window_channel.h [hunk 2]
```diff
@@ -0,0 +1,45 @@
+// Copyright 2021 Samsung Electronics Co., Ltd. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#ifndef EMBEDDER_WINDOW_CHANNEL_H_
+#define EMBEDDER_WINDOW_CHANNEL_H_
+
+#include <memory>
+#include <queue>
```
Comment (swift-kim) [inline shell/platform/tizen/channels/window_channel.h:9]: Unused include.

File: shell/platform/tizen/flutter_tizen_engine.cc [hunk 3]
```diff
@@ -394,20 +407,32 @@ void FlutterTizenEngine::SetWindowOrientation(int32_t degree) {
...
-                                  degree);
-    SendWindowMetrics(height, width, 0.0);
-  } else {
-    renderer_->ResizeWithRotation(geometry.x, geometry.y, width, height,
-                                  degree);
-    SendWindowMetrics(width, height, 0.0);
+    std::swap(width, height);
   }
+  renderer_->ResizeWithRotation(geometry.x, geometry.y, width, height, degree);
+  // Window position does not change on rotation regardless of it's size.
```
Comment (swift-kim) [inline shell/platform/tizen/flutter_tizen_engine.cc:413]: Maybe
```suggestion
  // Window position does not change on rotation regardless of its orientation.
```

File: shell/platform/tizen/flutter_tizen_engine.cc [hunk 4]
```diff
@@ -394,20 +407,32 @@ void FlutterTizenEngine::SetWindowOrientation(int32_t degree) {
...
+  SendWindowMetrics(geometry.x, geometry.y, width, height, 0.0);
 }
 
 void FlutterTizenEngine::OnOrientationChange(int32_t degree) {
   SetWindowOrientation(degree);
 }
 
+#ifndef TIZEN_RENDERER_EVAS_GL
```
Comment (swift-kim) [inline shell/platform/tizen/flutter_tizen_engine.cc:421]: Why don't you make `SetGeometry` as a member of `TizenRenderer` and remove this ifdef?

The body of `TizenRendererEvasGL::SetGeometry` could be just `FT_UNIMPLEMENTED()`.

File: shell/platform/tizen/flutter_tizen_engine.h [hunk 5]
```diff
@@ -215,6 +226,9 @@ class FlutterTizenEngine : public TizenRenderer::Delegate {
 #ifndef __X64_SHELL__
   // A plugin that implements Tizen app_control channels.
   std::unique_ptr<AppControlChannel> app_control_channel_;
+
+  // A plugin that implements the Tizen window channel.
+  std::unique_ptr<WindowChannel> window_channel_;
```
Comment (swift-kim) [inline shell/platform/tizen/flutter_tizen_engine.h:231]: Please sort the channels in alphabetical order.

Comment (HakkyuKim) [reply shell/platform/tizen/flutter_tizen_engine.h]: Done.

File: shell/platform/tizen/flutter_tizen_engine.cc (hunk 4 above)
Comment (HakkyuKim) [reply shell/platform/tizen/flutter_tizen_engine.cc]: Great! I could remove a whole lot of `#ifdef`s.

File: shell/platform/tizen/channels/window_channel.cc [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+
+namespace flutter {
+
+namespace {
+
+constexpr char kChannelName[] = "tizen/internal/window";
+
+}  // namespace
+
+#ifdef TIZEN_RENDERER_EVAS_GL
```
[the remark this answers is not in the document]
Comment (HakkyuKim) [reply shell/platform/tizen/channels/window_channel.cc:21]: This was necessary to suppress "unused variable `delegate`" error at compiling if the implementation used `#ifdef`, I'll instead check the render's dynamic type.

File: shell/platform/tizen/channels/window_channel.cc [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+                                width ? *width : geometry.w,
+                                height ? *height : geometry.h);
+    result->Success();
+#endif
+  } else if (method_name == "getScreenSize") {
+    int width, height;
+    system_info_get_platform_int("http://tizen.org/feature/screen.width",
+                                 &width);
+    system_info_get_platform_int("http://tizen.org/feature/screen.height",
+                                 &height);
```
[the remark this answers is not in the document]
Comment (HakkyuKim) [reply shell/platform/tizen/channels/window_channel.cc]: Must have been tricky bug to find, thanks!

File: shell/platform/tizen/channels/window_channel.cc (hunk 6 above)
[the remark this answers is not in the document]
Comment (HakkyuKim) [reply shell/platform/tizen/channels/window_channel.cc]: I wanted to implement a different solution based on the guide [here](https://google.github.io/styleguide/cppguide.html#Run-Time_Type_Information__RTTI_) but I couldn't think of a clean way to do it. I'll come back to it when I either implement the feature for evas or think of a design that would not require querying runtime type.

[the remark this answers is not in the document]
Comment (swift-kim) [reply shell/platform/tizen/channels/window_channel.cc]: Suppressing the unused variable compiler warning is relatively simple. You don't have to change the logic itself.

https://github.com/flutter-tizen/engine/blob/ad8d61f7339b4844cdfa92771fa8d6bf363e4bdd/shell/platform/tizen/external_texture_surface_gl_linux.cc#L5-L8

File: shell/platform/tizen/channels/window_channel.cc [hunk 8]
```diff
@@ -0,0 +1,80 @@
+// Copyright 2021 Samsung Electronics Co., Ltd. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "window_channel.h"
+
+#include <system_info.h>
```
Comment (swift-kim) [inline shell/platform/tizen/channels/window_channel.cc:7]: Unused include?

Comment (HakkyuKim) [reply shell/platform/tizen/channels/window_channel.cc:7]: Sorry, forgot to remove it after removing:

```
system_info_get_platform_int("http://tizen.org/feature/screen.width", &width);
system_info_get_platform_int("http://tizen.org/feature/screen.height", &height);
```

File: shell/platform/tizen/channels/window_channel.cc (hunk 6 above)
[the remark this answers is not in the document]
Comment (HakkyuKim) [reply shell/platform/tizen/channels/window_channel.cc]: Ah that's good, the changes are small if I apply that: https://github.com/HakkyuKim/engine/commit/eb0ae008f4ba3687abd4d760d982625ea9d9253e.

I'm slightly uncomfortable with exposing `OnGeometryChange` in `flutter_tizen_engine` when it doesn't work for `TizenRendererEvasGL` but I would have to add too many `#ifdef`s if I want to hide it, any suggestions?

[the remark this answers is not in the document]
Comment (swift-kim) [reply shell/platform/tizen/channels/window_channel.cc]: Hmm. Isn't this enough?

```cpp
void FlutterTizenEngine::OnGeometryChange(int32_t x,
                                          int32_t y,
                                          int32_t width,
                                          int32_t height) {
#ifdef TIZEN_RENDERER_EVAS_GL
  FT_UNIMPLEMENTED();
#else
  if (!renderer_->IsValid()) {
  ...
#endif
}
```

[the remark this answers is not in the document]
Comment (HakkyuKim) [reply shell/platform/tizen/channels/window_channel.cc]: Right, I didn't have to hide the declaration. Thanks for the suggestion!

File: shell/platform/tizen/flutter_tizen_engine.cc [hunk 9]
```diff
@@ -397,20 +405,33 @@ void FlutterTizenEngine::SetWindowOrientation(int32_t degree) {
...
   touch_event_handler_->rotation = degree;
   if (degree == 90 || degree == 270) {
-    renderer_->ResizeWithRotation(geometry.x, geometry.y, height, width,
-                                  degree);
-    SendWindowMetrics(height, width, 0.0);
-  } else {
-    renderer_->ResizeWithRotation(geometry.x, geometry.y, width, height,
-                                  degree);
-    SendWindowMetrics(width, height, 0.0);
+    std::swap(width, height);
```
Comment (bwikbs) [inline shell/platform/tizen/flutter_tizen_engine.cc:408]: Oh It's good!

**Comment by bwikbs** [comment]
Just curious... why did you disable at  x64? (Is this for evas issue? Same result on x64?)
I think it would be good to include a unit test..

**Comment by HakkyuKim** [comment]
> Just curious... why did you disable at x64? (Is this for evas issue? Same result on x64?) I think it would be good to include a unit test..

My understanding was that you could only test apis that are purely EFL on x64, the current implementation calls tizen only apis; `ecore_wl2_window_position_set` and `ecore_wl2_egl_window_resize_with_rotation`. Or am I mistaken?

File: shell/platform/tizen/channels/window_channel.cc [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wunused-private-field"
+#include "window_channel.h"
+#pragma clang diagnostic pop
+
+#include "flutter/shell/platform/common/client_wrapper/include/flutter/standard_method_codec.h"
+#include "flutter/shell/platform/tizen/channels/encodable_value_holder.h"
+#include "flutter/shell/platform/tizen/logger.h"
+namespace flutter {
```
Comment (swift-kim) [inline shell/platform/tizen/channels/window_channel.cc:13]:
```suggestion
#include "flutter/shell/platform/tizen/logger.h"

namespace flutter {
```

File: shell/platform/tizen/tizen_renderer_ecore_wl2.cc [hunk 11]
```diff
@@ -206,13 +206,19 @@ void* TizenRendererEcoreWl2::OnProcResolver(const char* name) {
...
-  WindowGeometry result;
+TizenRenderer::Geometry TizenRendererEcoreWl2::GetWindowGeometry() {
+  Geometry result;
   ecore_wl2_window_geometry_get(ecore_wl2_window_, &result.x, &result.y,
                                 &result.w, &result.h);
   return result;
 }
 
+TizenRenderer::Geometry TizenRendererEcoreWl2::GetScreenGeometry() {
+  Geometry result;
```
Comment (swift-kim) [inline shell/platform/tizen/tizen_renderer_ecore_wl2.cc:217]: Explicitly initialize with zeros, just to be safe.
```suggestion
  Geometry result = {};
```

**Review by swift-kim — APPROVED** [review]
Otherwise looks good to me!

**Comment by HakkyuKim** [comment]
> Otherwise looks good to me!

Done! Thanks for the review:)

**Review by bwikbs — APPROVED** [review]
> My understanding was that you could only test apis that are purely EFL on x64, the current implementation calls tizen only apis; ecore_wl2_window_position_set and ecore_wl2_egl_window_resize_with_rotation. Or am I mistaken?

First , it's not really important. I just asked.... 
I think I changed the window geometry in x64 before..  and we could only test the window channel operation (Not real behavior!) at unit test..  I just wondered why it was blocked at this time. Now I understand through your response.

Thanks! It's pretty cool! 👍

**Comment by bbrto21** [comment]
> @bbrto21 @xiaowei-guan Would you happen to have any information on this?

I'm working on another task else, so I can't afford to look into this. Sorry for the late reply.

**Comment by HakkyuKim** [comment]
> > My understanding was that you could only test apis that are purely EFL on x64, the current implementation calls tizen only apis; ecore_wl2_window_position_set and ecore_wl2_egl_window_resize_with_rotation. Or am I mistaken?
> 
> First , it's not really important. I just asked.... I think I changed the window geometry in x64 before.. and we could only test the window channel operation (Not real behavior!) at unit test.. I just wondered why it was blocked at this time. Now I understand through your response.
> 
> Thanks! It's pretty cool! +1

I was double checking if I was correct about the `__X64_SHELL__` define, I appreciate the review. I haven't planned adding unit test at the moment, maybe some other time.

**Comment by swift-kim** [comment]
Please resolve conflicts.

**Comment by HakkyuKim** [comment]
> Please resolve conflicts.

Done, merging would've been better, the conflict was:
```diff
-event.width = static_cast<size_t>(width);
-event.height = static_cast<size_t>(height);
+event.width = width;
+event.height = height;
```

**Pull request merged**